moonraker/plugins/history.py

```diff
@@ -0,0 +1,253 @@
+        args = web_request.get_args()
+        if "id" in args:
+            if args['id'] not in self.jobs:
+                raise self.server.error(f"Invalid job id: {args['id']}")
+            return {args['id']: self.get_job(args['id']).get_stats()}
+
+        before = None if 'before' not in args else int(args['before'])
+        since = None if 'since' not in args else int(args['since'])
+        limit = int(args['limit']) if "limit" in args else 50
+        start = int(args['start']) if "start" in args else 0
+        if start > (len(self.jobs)-1) or len(self.jobs) == 0:
+            return {"count": len(self.jobs), "prints": {}}
+
+        jobs = {}
+        i = 0
+        start_num = 0
+        end_num = len(self.jobs)
+        for id in list(self.jobs):
+            if since != None and since > self.get_job(id).get('start_time'):
+                start_num += 1
+                continue
+            if before != None and before < self.get_job(id).get('end_time'):
+                end_num -= 1
+                continue
+            if limit != 0 and i >= limit:
+                continue
+            if start != 0:
+                start -= 1
+                continue
+            jobs[id] = self.get_job(id).get_stats()
+            i += 1
+
+        return {"count": end_num - start_num, "prints": jobs}
+
+    async def _status_update(self, data):
+        if "print_stats" in data:
+            ps = data['print_stats']
+
+            if "filename" in ps:
+                file_store = self.server.lookup_plugin('file_manager')
+                if ps['filename'] != "":
+                    self.file_metadata = file_store.get_file_metadata(
+                        ps['filename'])
+                else:
+                    self.file_metadata = None
+
+            if "state" in ps:
+                old_state = self.print_stats['state']
+                new_state = ps['state']
+
+                if new_state is not old_state:
+                    if new_state == "printing" and old_state != "paused":
+                        for s in list(ps):
+                            self.print_stats[s] = ps[s]
+                        self.add_job(PrinterJob(self.print_stats))
+                    elif new_state == "complete" and self.current_job != None:
+                        for s in list(ps):
+                            self.print_stats[s] = ps[s]
+                        self.finish_job("completed", self.print_stats)
+                    elif new_state == "standby"  and self.current_job != None:
+                        self.finish_job("cancelled", self.print_stats)
+
+            for s in list(ps):
+                self.print_stats[s] = ps[s]
+
+            if time.time() > self.last_update_time + SAVE_INTERVAL:
+                if self.current_job != None:
+                    self.last_update_time = time.time()
+                    self.jobs[self.current_job].update_from_ps(self.print_stats)
+                    self.save_current_job()
```

Comment:
The saving interval is mainly in case of a power outage or some other un-foreseeable circumstance.  I can register to the other events, however.

Reply:
Ok, that makes sense.

Reply:
I still think it would be beneficial to save every so often for filament tracking purposes. While we may not have a spool manager yet, it would still be beneficial to know expended filament over time and track wasted vs useful filament.

Reply:
Understood, however I would like to discourage periodic updating to the database.  In the short term updating one record every 5 seconds isn't a big deal.  However if we have several plugins and/or clients updating on an interval we are inviting trouble. 

Consider the power loss scenario.  If the database is updating on an interval, it is will syncing at an interval as well.  Should the power loss occur during a sync it is possible to corrupt the database.  If it occurs while lmdb is writing to disk it could corrupt the entire file system.

Reply:
FWIW, this particular issue won't keep me from merging.  I think the chances of a system fault, such as a power outage, during a sync are very small.